src/main/java/com/uid2/shared/attest/AttestationTokenRetriever.java

```diff
@@ -0,0 +1,201 @@
+package com.uid2.shared.attest;
+
+import com.uid2.enclave.AttestationException;
+import com.uid2.enclave.IAttestationProvider;
+import com.uid2.shared.*;
+import io.vertx.core.Handler;
+import io.vertx.core.Vertx;
+import io.vertx.core.json.Json;
+import io.vertx.core.json.JsonObject;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.net.*;
+import java.net.http.HttpClient;
+import java.net.http.HttpRequest;
+import java.net.http.HttpResponse;
+import java.nio.charset.StandardCharsets;
+import java.security.*;
+import java.time.Instant;
+import java.util.Base64;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
+public class AttestationTokenRetriever {
+    private static final Logger LOGGER = LoggerFactory.getLogger(AttestationTokenRetriever.class);
+    private final IAttestationProvider attestationProvider;
+    private final ApplicationVersion appVersion;
+    private AtomicReference<String> attestationToken;
+    private Handler<Integer> responseWatcher;
+    private final String attestationEndpoint;
+    private final HttpClient httpClient;
+    private final IClock clock;
+    private Vertx vertx;
+    private boolean isExpiryCheckScheduled;
+    // Set this to be Instant.MAX so that if it's not set it won't trigger the re-attest
+    private Instant attestationTokenExpiresAt = Instant.MAX;
+    private final Lock lock;
+    private final AttestationTokenDecryptor attestationTokenDecryptor;
+
+    public AttestationTokenRetriever(String attestationEndpoint, ApplicationVersion appVersion, IAttestationProvider attestationProvider,
+                                     Handler<Integer> responseWatcher, IClock clock, HttpClient httpClient,
+                                     AttestationTokenDecryptor attestationTokenDecryptor) throws IOException {
+        this.attestationEndpoint = attestationEndpoint;
+        this.appVersion = appVersion;
+        this.attestationProvider = attestationProvider;
+        this.attestationToken = new AtomicReference<>(null);
+        this.responseWatcher = responseWatcher;
+        this.clock = clock;
+        this.lock = new ReentrantLock();
+        this.isExpiryCheckScheduled = false;
+        if (httpClient == null) {
+            this.httpClient = HttpClient.newHttpClient();
+        } else {
+            this.httpClient = httpClient;
+        }
+        if (attestationTokenDecryptor == null) {
+            this.attestationTokenDecryptor = new AttestationTokenDecryptor();
+        } else {
+            this.attestationTokenDecryptor = attestationTokenDecryptor;
+        }
+    }
+
+    private void attestationExpirationCheck(long timerId) {
+        Instant currentTime = clock.now();
```

Comment:
Now that you don't un-schedule it, what happens if the attest takes longer than 60sec? This will get called again, and will start another attest.
I suggest we add a flag around this code, so that if we are still in the process of attest, when this gets called again, it just exists without trying to attest again. We should also log that we have skipped the attest so we can detect if that is happening too much